src/CommonRuntime/IO.fs

```diff
@@ -192,7 +201,11 @@ let private watchForChanges (uri:Uri) (((tp:IDisposableTypeProvider), typeName)
 
         if (match typeNameBeingDisposedOpt with None -> true | Some typeNameBeingDisposed -> typeName = typeNameBeingDisposed) then 
             if watcher.Remove tp typeName then
-                watchers.Remove uri.OriginalString |> ignore
+#if FX_NO_CONCURRENT 
+                lock watchers <| fun () -> watchers.Remove uri.OriginalString |> ignore
+#else 
+                watchers.TryRemove(uri.OriginalString) |> ignore
+#endif
```

Comment:
There's a potential threading issue here. The if watcher.Remove needed to be done inside the lock. To fix it we have to keep using explicit lock, I'll push the fix to your branch before merging

Reply:
done